### ISSUE TITLE
Packet 255, ViewArea

### ISSUE THREAD

**Issue by terjanq** [SNSA/agario-client2#13]
[255,33,0,0,0,240,18,64,MinX::Float64, MnY::Float64, MaxX::Float64, MaxY::Float64]

255 - packet ID
33,0,0,0 -> Uint32 -> packet length
240, 18 -> SubId for ViewArea

if(MaxX - MinX > 14E3 && MaxY - MinY > 14E3) game.offsetX = 7071 + Minx, game.offsetY = 7071 + MaxY

**Comment by heroandtn3** [comment]
According to [old protocol](http://agar.gcommer.com/index.php?title=Protocol#World_Update_.28opcode_16.29), I guess `33,0` is number of blob eat events.

**Comment by terjanq** [comment]
I told what it is, it is old 64 packet.

33,0,0,0 - packet length Uint32

18 = 33-15, 

240, ..., 64-> looks like old protocold ID.

**Comment by heroandtn3** [comment]
I think all fields should have 1, 2, or 4 bytes, while 240, 18, 64 is 3 bytes so it's invalid IMO.

Btw, this is what I got from @q3wzeck:

```JavaScript
255: (data) => {
        //logger.debug('World update:', data);
        var buf = new Buffer(data);
        var packet    = new Packet(buf);
        var packet_id = packet.readUInt8();
        var packet_subId = packet.readUInt8();

        if (packet_subId == 43) {

          packet.offset += 6;
          var min_x = packet.readFloat64LE();
          var min_y = packet.readFloat64LE();
          var max_x = packet.readFloat64LE();
          var max_y = packet.readFloat64LE();

          this.emit('mapSize', min_x, min_y, max_x, max_y);

        };


        this.emit('worldUpdate', data);
      }
```

**Comment by terjanq** [comment]
It's wrong.

It is first packet arrived, it cointans also server version at the end of it.

My schema for packets is 100% correct:

[255:Uint8, 33::Uint32, 240::Uint8, 18::Uint32, 64::Uint8, MinX::Float64, MinY::Float64, MaxX::Float64, MaxY::Float64]

You need detect 240 packet which means "Protcol will be loaded", I don't know what 18 is for, but as i noticed it is 33-15 (repeatable pattern), and 64 it is field which infrom you it is ViewArea.

**Comment by heroandtn3** [comment]
Oh, can you send a PR for packet 255? I think it would not be hard, like this one: https://github.com/SNSA/agario-client2/blob/master/src/PacketDecoder.js

**Comment by hassanila97** [comment]
packet 255 includes packet 16 and more....

**Comment by hassanila97** [comment]
This is all you need: https://github.com/xzfc/agar-expose/blob/master/expose.user.js

**Comment by q3wzeck** [comment]
@terjanq its not wrong
`mapSizeChanged: min_x: -6342.517032679486 min_y: -5093.496457163584 max_x: 7799.618591051466 max_y: 9048.63916656737`
but 2nd packet with coordinates sends other dimesions.
and there is no more 33 subpacket. i dont see it in 255 now.

**Comment by terjanq** [comment]
@q3wzeck  Of course it is wrong. 33 It's not subID, also It's not Uint8, its Uint32.  packet


https://sc-cdn.scaleengine.net/i/4d16f9f848d2f59f2ee4230e4b839cf7.png

**Comment by q3wzeck** [comment]
@terjanq omg. ok.
what software u use? O_O

**Comment by terjanq** [comment]
I've made it :)

**Comment by hassanila97** [comment]
Any1 has moveto packet?

**Comment by terjanq** [comment]
I have  decoded 90% already

**Comment by q3wzeck** [comment]
@hassanila97 it still has id 16, but need key in last 4 bytes.

**Comment by hassanila97** [comment]
Mhh where do I get the key?

**Comment by q3wzeck** [comment]
@hassanila97 its generating llvm inside agario.core.js, so idk :(

**Comment by terjanq** [comment]
@hassanila97 It is function in my core.js which generate every packet.

http://wklej.org/hash/368700089aa/

**Comment by hassanila97** [comment]
@terjanq Yes, but it is called with 2 arguments and based on them it generates the keys

**Comment by q3wzeck** [comment]
@terjanq its generating 113 and 16?

**Comment by terjanq** [comment]
Don't remember if 113 also, i was trying it days ago. It's generating 16 packet. 

Packet 16 depends of first 32 packet (cell ID) after packet 18(reset cells data) has arrived.

**Comment by rebrec** [comment]
>  terjanq commented 7 hours ago • edited
> 
> It's wrong.
> 
> It is first packet arrived, it cointans also server version at the end of it.
> 
> My schema for packets is 100% correct:
> 
> [255:Uint8, 33::Uint32, 240::Uint8, 18::Uint32, 64::Uint8, MinX::Float64, MinY::Float64, MaxX::Float64, MaxY::Float64]
> 
> All you need is detect 240 field which means "Protcol will be loaded", I don't know what 18 is for, but as i noticed it is 33-15 (repeatable pattern), and 64 it is field which infrom you it is ViewArea.
> 

Just a little mistake to mention :  [255:Uint8, 33::Uint32, 240::Uint8, 18::Uint**8 (not 32)**, 64::Uint8, MinX::Float64, MinY::Float64, MaxX::Float64, MaxY::Float64]

Thanks for this decoding

**Comment by terjanq** [comment]
Yeah, of course. Also according with their schema there could be other data inside it (before and after viewarea) but actually i have noticed only the first packet which contains server version as well.

**Comment by terjanq** [comment]
This is my tool for decoding. 

http://agarpp.comli.com/analyse/

Also look here: https://github.com/OgarProject/Ogar/issues/569#issuecomment-220280584

**Comment by Barbosik** [comment]
They released protocol 7 which uses encrypted communication :(

**Comment by hassanila97** [comment]
Fuck that shit

**Comment by Barbosik** [comment]
Yeah, I had a hope to play in my favorite game again, but now it's complete shit.

**Comment by terjanq** [comment]
I think they have prepared many patches and when they see bots are back, they use next one... Bots has gone again tho

**Comment by Barbosik** [comment]
The good thing is that it doesn't affect private servers. So, with fix for protocol 6 in Ogar server it works ok on protocol 7 client.
They just added encryption over protocol 6, and server can activate it by sending encryption key to the client.

**Comment by terjanq** [comment]
I have updated some raw packets on my site

**Comment by Barbosik** [comment]
**terjanq** What site you're talking about?

Do u know good private servers?

**Comment by terjanq** [comment]
http://agarpp.comli.com/analyse/

This tool i meant

**Comment by Barbosik** [comment]
For what? The protocol is uncovered and all things is pretty clear, except some nuances with compression. But now server encrypt all messages. You can read only first messages. And later server sends encryption key and switch to encrypted messages.
In order to read it there is need to decode keys from agar.core, it's a crap. 
Only bot owners with moneys can afford it.

**Comment by Barbosik** [comment]
I was wrong, all messages from the the client in protocol 7 is encrypted.
So, you will not be able to receive any command from the client in Ogar server.

**Comment by terjanq** [comment]
For better see packets and for fun. Also it easier finding part of code which i need.

**Comment by Barbosik** [comment]
It's better to use notepad and editor in the FAR manager for such purposes.

By the way, here is a good news, they removed encryption for private servers. It's still used for official servers to prevent bots. But for private servers encryption is not used now. So we can revive private servers.

You can find details about the new protocol for private servers here: https://github.com/issy123/agario-protocol/issues/15
